src/server/game/Server/WorldSession.cpp

```diff
@@ -847,8 +847,20 @@ void WorldSession::ReadMovementInfo(WorldPacket &data, MovementInfo* mi)
         in conjunction with any of the moving movement flags such as MOVEMENTFLAG_FORWARD.
         It will freeze clients that receive this player's movement info.
     */
-    REMOVE_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_ROOT),
-        MOVEMENTFLAG_ROOT);
+    if (!GetPlayer()->GetVehicleBase())
+    {
+        // Only remove here for non vehicles
+        REMOVE_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_ROOT),
+            MOVEMENTFLAG_ROOT);
+    }
+    else
+    {
+        // Actually players in rooted vehicles still send commands, don't clear root for these!
+        // Check for non movement flag. Otherwise, also clear for vehicles
+        if (!GetPlayer()->GetVehicleBase()->HasFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_DISABLE_MOVE))
+            REMOVE_VIOLATING_FLAGS(mi->HasMovementFlag(MOVEMENTFLAG_ROOT),
+                MOVEMENTFLAG_ROOT);
```

Comment:
Actually expanding on this. From what I can see, the code rather than blocking clients sending movement packets when rooted, releases the root. Which is bad. Probably if the packets were blocked, the offending client would "move" but all other players wouldn't see them move and the server would put them in their original place. Isn't this a better overall solution?

In any case, even if you did this. Other players wouldn't see the turrets of the cannons moving. I'm pretty sure this would be equally wrong. So, I think it has to be allowed here. I'm not sure how easy it would be to isolate axis movement rather than location movement here though.

That would be the 100% "correct" way to do it and should prevent any spoofing freezes.

Would you agree?